CORDA-3994 Retry errors in flow init started with client ids

lankydan opened the pull request:
Flows that were started with a client id would hang because it would
retrieve the existing flow's future and wait for it to finish. But,
because the flow has failed its flow init and not saved its initial
checkpoint, it is relying on `startFlow` to start the flow again (by
redelivering the start flow external event).

`FlowWithClientIdStatus` now holds the flow id that it is related to.
This is then checked in `startFlow`. If a matching client id is found
for a flow start, it then checks the flow id as well. If the flow id
matches, then it lets the `startFlow` call continue, allowing it to
actually start the flow again (how a flow without a client id would
retry in this situation).

lankydan commented:
Run tests: slow

chrisr3 commented:
run tests: smoke, slow

lankydan commented:
run tests: slow

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -311,17 +312,20 @@ internal class SingleThreadedStateMachineManager(
-                return@startFlow uncheckedCast(existingFuture)
-            }
-            onClientIDNotFound?.invoke()
+                // Do not exit if the flow id from the external event matches the existing flow
```
I think we can improve the comment here.  I found it a little confusing (exit was odd choice of word).

`If the flow ID is the same as the one recorded in the client ID map, then this start flow event has been retried, and we should not de-duplicate.`

lankydan committed:
CORDA-3994 Retry errors in flow init started with client ids

Flows that were started with a client id would hang because it would
retrieve the existing flow's future and wait for it to finish. But,
because the flow has failed its flow init and not saved its initial
checkpoint, it is relying on `startFlow` to start the flow again (by
redelivering the start flow external event).

`FlowWithClientIdStatus` now holds the flow id that it is related to.
This is then checked in `startFlow`. If a matching client id is found
for a flow start, it then checks the flow id as well. If the flow id
matches, then it lets the `startFlow` call continue, allowing it to
actually start the flow again (how a flow without a client id would
retry in this situation).

williamvigorr3 reviewed (commented):
LGTM
